### DISCUSSION TITLE
Cache Busting for Static Assets

### DISCUSSION THREAD

**Discussion by mdmontesinos** [oqtane/oqtane.framework#4745] [General] [answered]
When deploying new versions of modules in a production environment, you might end up with updated code but obsolete css or js due to browser cache (at least in static mode), which is really problematic as the user will perceive a broken site (i.e major css changes or new sections whose css is not served).

Does Oqtane have any [cache busting](https://www.keycdn.com/support/what-is-cache-busting) system or something similar to prevent this issue? Usual techniques like bundlers do not work properly here because you need to reference the static assets in the ModuleInfo. It would be nice if Oqtane had some way to achieve this easily.

For example, on build time (perhaps only on release mode?), computing the hash of static assets and appending it to the resource declaration. In this way, when deploying an updated module, the hash of the modified assets would change and the browser would request them as if they were new. This effectively keeps the unmodifed files cached, and requests the modified ones to avoid the "broken" site state.

**Comment by sbwalker** [comment, marked as answer]
As a developer of a module, you obviously know when one of your static dependencies has  changed ie. a CSS or JavaScript file. If you want to ensure that it gets reloaded you can append a querystring to the Resource reference in your code:

```
            Resources = new List<Resource>()
            {
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/Module.css?v=12345" }
            }
```

This approach works well and is much more efficient than imposing a large overheard on the framework itself to compute and manage versions for every static asset - including those that rarely/never change.

**Reply by mdmontesinos** [reply]
You're right that perhaps it's too much overkill to have an in-framework automated solution.

Thanks for your insights!

**Comment by nicpitsch** [comment]
@sbwalker will it be possible to use the new capabilites of .NET 9 like Fingerprinting (presented here: https://youtu.be/o0CWssf8TFw?si=43WJp4HDqoREiIPl&t=2550) in Oqtane respectively Oqtane modules?

**Reply by mdmontesinos** [reply]
It would be really nice if Oqtane could support the new MapStaticAssets instead of UseStaticFiles to get out-of-the-box compression and fingerprinting of static assets (https://learn.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-9.0?view=aspnetcore-8.0#optimizing-static-web-asset-delivery)

Also, the new [HybridCache](https://learn.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-9.0?view=aspnetcore-8.0#new-hybridcache-library) is a must, as it is a drop-in replacement for the currently used IMemoryCache and prevents cache reentrancy (or stampede as they call it in the docs)

**Reply by sbwalker** [reply]
Yes there are a number of items which Oqtane should be able to take advantage of in . NET 9. HybridCache in particular will be used as part of the Scale Out enhancement on the roadmap.
